src/hooks/usePersistentQuery.ts

```diff
@@ -0,0 +1,36 @@
+import { useNavigate } from 'react-router-dom';
+import { getBasePath } from 'utils/formatPath';
+import useQueryParams from 'hooks/useQueryParams';
+import { usePersistentGlobalState } from './usePersistentGlobalState';
+import { useCallback } from 'react';
+
+export const usePersistentQuery = <T extends Record<string, string>>(
+    key: string
+) => {
+    const params = useQueryParams();
+    const navigate = useNavigate();
+    const { data, mutate } = usePersistentGlobalState(
+        `${getBasePath()}:${key}:v1`,
+        Object.fromEntries(params)
+    );
+
+    const setQuery = useCallback<(newData: T, replace?: boolean) => void>(
+        (newData, replace = true) => {
+            if (
+                Object.keys(newData).length === Object.keys(data).length &&
```

Comment:
Could be extracted as a helper function for clarity `if (matchesQuery(newData, data)`